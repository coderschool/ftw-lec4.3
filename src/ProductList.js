import React from 'react'

const ProductList = ({products, onProductBuy}) => {
  return (
    <div>
      { products.map( p => <li onClick={() => onProductBuy(p.id)}>{p.name}: {p.price}<button> add to cart </button><p>Weird Extra Code</p></li>)}
    </div>
  )
}

export default ProductList
