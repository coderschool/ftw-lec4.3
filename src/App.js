import React, { Component } from 'react';
import logo from './logo.svg';
import './App.css';
import ProductList from './ProductList';

class App extends Component {
  constructor() {
    super();
    this.state = {
      products: []
    }
  }
  render() {
    return (
      <div className="App">
        <h1>Charles-Zada</h1>
        <ProductList products={this.state.products} />
      </div>
    );
  }
}

export default App;
