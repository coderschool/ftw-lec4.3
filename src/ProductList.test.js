import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import { shallow } from 'enzyme';
import ProductList from './ProductList';

let TEST_DATA;
let output;
let testingFn;

beforeEach( () => {
  testingFn = jest.fn();
  TEST_DATA = [
    {id: 1, name: 'computer', price: 1000},
    {id: 2, name: 'anker mouse', price: 50},
    {id: 3, name: 'computer stand', price: 20}
  ];
   output = shallow(
   <ProductList 
    onProductBuy={testingFn} 
    products={TEST_DATA}/>);
});


it('renders products on the page', () => { 
  expect(output.find('li').length).toEqual(3);
});

it('renders the price of the product', () => { 
  const firstProduct = output.find('li').first(); 
  expect(firstProduct.text()).toMatch(/1000/);
});

it('calls onProductBuy when the element is clicked', () => {
  const firstProduct = output.find('li').first(); 
  // First, check that nobody has called our function yet.
  expect(testingFn.mock.calls.length).toEqual(0);
  firstProduct.simulate('click');
  expect(testingFn.mock.calls.length).toEqual(1);
  // Check the first call to this function has the argument of 1
  expect(testingFn.mock.calls[0][0]).toEqual(1);
});

it('has one button', () => {
  const firstProduct = output.find('li').first(); 
  const button = expect(firstProduct.find('button').length).toBe(1);
});

it('matches our snapshot', () => {
  expect(output).toMatchSnapshot();
});
