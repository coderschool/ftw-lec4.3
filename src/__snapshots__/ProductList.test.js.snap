// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches our snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ProductList
    onProductBuy={[MockFunction]}
    products={
      Array [
        Object {
          "id": 1,
          "name": "computer",
          "price": 1000,
        },
        Object {
          "id": 2,
          "name": "anker mouse",
          "price": 50,
        },
        Object {
          "id": 3,
          "name": "computer stand",
          "price": 20,
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <li
          onClick={[Function]}
        >
          computer
          : 
          1000
          <button>
             add to cart 
          </button>
          <p>
            Weird Extra Code
          </p>
        </li>,
        <li
          onClick={[Function]}
        >
          anker mouse
          : 
          50
          <button>
             add to cart 
          </button>
          <p>
            Weird Extra Code
          </p>
        </li>,
        <li
          onClick={[Function]}
        >
          computer stand
          : 
          20
          <button>
             add to cart 
          </button>
          <p>
            Weird Extra Code
          </p>
        </li>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "computer",
            ": ",
            1000,
            <button>
               add to cart 
            </button>,
            <p>
              Weird Extra Code
            </p>,
          ],
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Array [
          "computer",
          ": ",
          1000,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": " add to cart ",
            },
            "ref": null,
            "rendered": " add to cart ",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Weird Extra Code",
            },
            "ref": null,
            "rendered": "Weird Extra Code",
            "type": "p",
          },
        ],
        "type": "li",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "anker mouse",
            ": ",
            50,
            <button>
               add to cart 
            </button>,
            <p>
              Weird Extra Code
            </p>,
          ],
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Array [
          "anker mouse",
          ": ",
          50,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": " add to cart ",
            },
            "ref": null,
            "rendered": " add to cart ",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Weird Extra Code",
            },
            "ref": null,
            "rendered": "Weird Extra Code",
            "type": "p",
          },
        ],
        "type": "li",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "computer stand",
            ": ",
            20,
            <button>
               add to cart 
            </button>,
            <p>
              Weird Extra Code
            </p>,
          ],
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Array [
          "computer stand",
          ": ",
          20,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": " add to cart ",
            },
            "ref": null,
            "rendered": " add to cart ",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Weird Extra Code",
            },
            "ref": null,
            "rendered": "Weird Extra Code",
            "type": "p",
          },
        ],
        "type": "li",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <li
            onClick={[Function]}
          >
            computer
            : 
            1000
            <button>
               add to cart 
            </button>
            <p>
              Weird Extra Code
            </p>
          </li>,
          <li
            onClick={[Function]}
          >
            anker mouse
            : 
            50
            <button>
               add to cart 
            </button>
            <p>
              Weird Extra Code
            </p>
          </li>,
          <li
            onClick={[Function]}
          >
            computer stand
            : 
            20
            <button>
               add to cart 
            </button>
            <p>
              Weird Extra Code
            </p>
          </li>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "computer",
              ": ",
              1000,
              <button>
                 add to cart 
              </button>,
              <p>
                Weird Extra Code
              </p>,
            ],
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Array [
            "computer",
            ": ",
            1000,
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": " add to cart ",
              },
              "ref": null,
              "rendered": " add to cart ",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Weird Extra Code",
              },
              "ref": null,
              "rendered": "Weird Extra Code",
              "type": "p",
            },
          ],
          "type": "li",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "anker mouse",
              ": ",
              50,
              <button>
                 add to cart 
              </button>,
              <p>
                Weird Extra Code
              </p>,
            ],
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Array [
            "anker mouse",
            ": ",
            50,
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": " add to cart ",
              },
              "ref": null,
              "rendered": " add to cart ",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Weird Extra Code",
              },
              "ref": null,
              "rendered": "Weird Extra Code",
              "type": "p",
            },
          ],
          "type": "li",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "computer stand",
              ": ",
              20,
              <button>
                 add to cart 
              </button>,
              <p>
                Weird Extra Code
              </p>,
            ],
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Array [
            "computer stand",
            ": ",
            20,
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": " add to cart ",
              },
              "ref": null,
              "rendered": " add to cart ",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Weird Extra Code",
              },
              "ref": null,
              "rendered": "Weird Extra Code",
              "type": "p",
            },
          ],
          "type": "li",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
